File: cpp/problems/leetcode/0010_regular_expression_matching/solution.cpp
// LeetCode 10
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        int m = (int)s.size(), n = (int)p.size();
        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
        dp[0][0] = true;
        for (int j = 2; j <= n; ++j)
            if (p[j - 1] == '*') dp[0][j] = dp[0][j - 2];
        for (int i = 1; i <= m; ++i)
            for (int j = 1; j <= n; ++j) {
                if (p[j - 1] == '*') {
                    dp[i][j] = dp[i][j - 2];
                    if (p[j - 2] == '.' || p[j - 2] == s[i - 1])
                        dp[i][j] = dp[i][j] || dp[i - 1][j];
                } else if (p[j - 1] == '.' || p[j - 1] == s[i - 1]) {
                    dp[i][j] = dp[i - 1][j - 1];
                }
            }
        return dp[m][n];
    }
};

int main() {
    assert(!Solution().isMatch("aa", "a"));
    assert(Solution().isMatch("aa", "a*"));
    assert(Solution().isMatch("ab", ".*"));
    assert(Solution().isMatch("aab", "c*a*b"));
    assert(!Solution().isMatch("mississippi", "mis*is*p*."));
    cout << "0010 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0010_regular_expression_matching/solution.py
"""LeetCode 10: Regular Expression Matching."""

from __future__ import annotations


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[0][0] = True
        for j in range(2, n + 1):
            if p[j - 1] == "*":
                dp[0][j] = dp[0][j - 2]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[j - 1] == "*":
                    dp[i][j] = dp[i][j - 2]
                    if p[j - 2] == "." or p[j - 2] == s[i - 1]:
                        dp[i][j] = dp[i][j] or dp[i - 1][j]
                else:
                    if p[j - 1] == "." or p[j - 1] == s[i - 1]:
                        dp[i][j] = dp[i - 1][j - 1]
        return dp[m][n]


if __name__ == "__main__":
    assert not Solution().isMatch("aa", "a")
    assert Solution().isMatch("aa", "a*")
    assert Solution().isMatch("ab", ".*")
    assert Solution().isMatch("aab", "c*a*b")
    assert not Solution().isMatch("mississippi", "mis*is*p*.")
    print("0010 OK")
